src/main/java/gregtech/common/blocks/BlockFrame.java

```diff
@@ -182,67 +140,68 @@ public boolean replaceWithFramedPipe(World worldIn, BlockPos pos, IBlockState st
-        ItemStack stackInHand = playerIn.getHeldItem(hand);
-        if (stackInHand.isEmpty()) {
+    public boolean onBlockActivated(@Nonnull World world, @Nonnull BlockPos pos, @Nonnull IBlockState state,
+                                    @Nonnull EntityPlayer player, @Nonnull EnumHand hand, @Nonnull EnumFacing facing,
+                                    float hitX, float hitY, float hitZ) {
+        ItemStack stack = player.getHeldItem(hand);
+        if (stack.isEmpty()) {
             return false;
         }
         // replace frame with pipe and set the frame material to this frame
-        if (stackInHand.getItem() instanceof ItemBlockPipe) {
-            return replaceWithFramedPipe(worldIn, pos, state, playerIn, stackInHand, facing);
+        if (stack.getItem() instanceof ItemBlockPipe) {
+            return replaceWithFramedPipe(world, pos, state, player, stack, facing);
         }
 
-        if (stackInHand.getItem().getToolClasses(stackInHand).contains(ToolClasses.CROWBAR)) {
-            return removeFrame(worldIn, pos, playerIn, stackInHand);
+        if (stack.getItem().getToolClasses(stack).contains(ToolClasses.CROWBAR)) {
+            return removeFrame(world, pos, player, stack);
         }
 
-        if (!(stackInHand.getItem() instanceof FrameItemBlock)) {
-            return false;
-        }
+        BlockFrame frameBlock = getFrameBlockFromItem(stack);
+        if (frameBlock == null) return false;
+
         BlockPos.PooledMutableBlockPos blockPos = BlockPos.PooledMutableBlockPos.retain();
         blockPos.setPos(pos);
         for (int i = 0; i < 32; i++) {
-            if (worldIn.getBlockState(blockPos).getBlock() instanceof BlockFrame) {
+            if (world.getBlockState(blockPos).getBlock() instanceof BlockFrame) {
                 blockPos.move(EnumFacing.UP);
                 continue;
             }
-            TileEntity te = worldIn.getTileEntity(blockPos);
+            TileEntity te = world.getTileEntity(blockPos);
             if (te instanceof IPipeTile && ((IPipeTile<?, ?>) te).getFrameMaterial() != null) {
                 blockPos.move(EnumFacing.UP);
                 continue;
             }
-            if (canPlaceBlockAt(worldIn, blockPos)) {
-                worldIn.setBlockState(blockPos, ((FrameItemBlock) stackInHand.getItem()).getBlockState(stackInHand));
-                SoundType type = getSoundType(stackInHand);
-                worldIn.playSound(null, pos, type.getPlaceSound(), SoundCategory.BLOCKS, (type.getVolume() + 1.0F) / 2.0F, type.getPitch() * 0.8F);
-                if (!playerIn.capabilities.isCreativeMode) {
-                    stackInHand.shrink(1);
+            if (canPlaceBlockAt(world, blockPos)) {
+                world.setBlockState(blockPos, this.getStateFromMeta(stack.getItem().getMetadata(stack.getItemDamage())));
```

Comment:
Can this use `stack.getMetadata()` instead?

Reply:
`getMetadata(int)` and `getMetadata(ItemStack)` are separate method which could have different implementation, like in base `Item` class.

Although changing to ItemStack one won't change the behavior, using int one is technically correct.